satpy/composites/__init__.py

```diff
@@ -1714,24 +1726,51 @@ def _combine_metadata_with_mode_and_sensor(self,
 
     @staticmethod
     def _get_merged_image_data(foreground: xr.DataArray,
-                               background: xr.DataArray
+                               background: xr.DataArray,
+                               initial_bg_alpha: bool,
                                ) -> list[xr.DataArray]:
-        if "A" in foreground.attrs["mode"]:
-            # Use alpha channel as weight and blend the two composites
-            alpha = foreground.sel(bands="A")
-            data = []
-            # NOTE: there's no alpha band in the output image, it will
-            # be added by the data writer
-            for band in foreground.mode[:-1]:
-                fg_band = foreground.sel(bands=band)
-                bg_band = background.sel(bands=band)
-                chan = (fg_band * alpha + bg_band * (1 - alpha))
-                chan = xr.where(chan.isnull(), bg_band, chan)
-                data.append(chan)
-        else:
-            data_arr = xr.where(foreground.isnull(), background, foreground)
-            # Split to separate bands so the mode is correct
-            data = [data_arr.sel(bands=b) for b in data_arr["bands"]]
+        def _get_alpha(dataset: xr.DataArray):
+            # If the dataset contains an alpha channel, just use it
+            # If not, we still need one. So build it and fill it with 1
+            if "A" in dataset.attrs["mode"]:
+                alpha = dataset.sel(bands="A")
+            else:
+                first_band = dataset.isel(bands=0)
+                alpha = xr.full_like(first_band, 1)
+                alpha["bands"] = "A"
+
+            # There could be Nans in the alpha, especially through 'add_bands'
+            # Replace them with 0 to prevent cases like 1 + nan = nan, so they won't affect new_alpha
+            alpha = xr.where(alpha.isnull(), 0, alpha)
+
+            return alpha
+
+        alpha_fore = _get_alpha(foreground)
+        alpha_back = _get_alpha(background)
+        new_alpha = alpha_fore + alpha_back * (1 - alpha_fore)
+
+        data = []
+
+        # Unless background has an initial alpha band, there will be no alpha band in the output image
+        # Let the writer decide
+        output_mode = background.attrs["mode"] if initial_bg_alpha else background.attrs["mode"].replace("A", "")
+
+        # If we let the writer decide alpha band, we must fill the transparent areas in the image with np.nan first
+        # The best way is through the new alpha
+        new_alpha_nan = xr.where(alpha_fore + alpha_back == 0, np.nan, new_alpha) if "A" not in output_mode \
+            else new_alpha
+
+        for band in output_mode:
+            fg_band = foreground.sel(bands=band) if band != "A" else new_alpha
+            bg_band = background.sel(bands=band) if band != "A" else new_alpha
+
+            chan = (fg_band * alpha_fore +
+                    bg_band * alpha_back * (1 - alpha_fore)) / new_alpha_nan if band != "A" else new_alpha
```

Comment:
What is the `/ new_alpha_nan` doing here?

Reply:
Like the comments say, when there's no initial alpha band in the background then we won't pass the alpha band to the writer. Instead we leave the writer for decision. Before that, we must fill all the transparent areas with nan. The best way to do this is letting the `new_alpha` get involved in `chan = (fg_band * alpha_fore + bg_band * alpha_back * (1 - alpha_fore)) / new_alpha`  ([https://en.wikipedia.org/wiki/Alpha_compositing](url)) since `anything / nan = nan`. However, I don't want to change the `new_alpha` so I make a duplicate here `new_alpha_nan`

Reply:
Thanks for the wikipedia article. I was starting to wonder if this was equivalent to the pre-multiplied alpha, but the article says it is a straight alpha so OK.

Would it be possible to separate out the alpha channel handling (the `band == "A"` case) to be done after the loop? Because in that case `chan` just becomes `new_alpha`, right? I'm not sure I understand what the below `xr.where(chan.isnull(), ...` does in that case though.

Next question, how much extra work is this compositor now doing in the case where no alphas are involved? In the old code there was the `else:` block that did a simple `xr.where`. Is it worth maybe have a special branch of `if "A" not in foreground_mode and "A" not in background_mode: return self._simple_overlay()` or something?

Reply:
1. Agree. Leaving the alpha outside the loop will make it more tidy.
2. `chan = xr.where(chan.isnull(), bg_band * alpha_back, chan)` is to fill the areas where `foreground` is completely transparent with `background`. Without it you may get an image like this (HighLowCloud_with_background):
![image](https://github.com/pytroll/satpy/assets/72339781/f282771d-2afb-47ef-9343-1fdcc46097fa)
Actually, it is:
![image](https://github.com/pytroll/satpy/assets/72339781/1384b0db-c725-4e20-880f-40880e9fa4c0)
3. Well I use this loop just to avoid `if/else` block and to make things more general. But if you want to simplify the workflow for no alpha cases I can bring it back(could be a little huge) or make another method for this just like you said.

Reply:
For 2, what about the data in this case is making it NaN? Why aren't other areas over land NaN in the cloud image? I guess another way for me to think about this is that the alpha band would have to be set to 0 (for foreground and background) where ever the band data is NaN, but also the band data would need to be set to 0 where ever the alpha is 0. Ok ok, too many cases to work around.

For 3, although the code got longer I do think overall it simplifies the understanding.

I think my biggest confusion with this code overall is that the NaN handling seems relatively magical. I mean, it isn't explicit and comes from some dividing and 0 or NaN filling at various parts. I might be the only one that is having trouble with this though so we should wait to see what @mraspaud and @pnuu say when they get back from their holidays.

Reply:
> but also the band data would need to be set to 0 where ever the alpha is 0

It could be a little risky for float32 output...So in my understanding, the band data should to be set to NaN. And that is exactly what `new_alpha_nan` does

Reply:
I understand your concern. It's kinda tricky here so we may wait for their thoughts.

Reply:
@mraspaud @pnuu I'd still like your comments here if you have time.

Reply:
Ok, I've been looking at the code here, and although I understand what the code is doing, I understand @djhoese 's concern regarding the nan magic. I think though it can be made clearer if these lines:
```
            chan = (fg_band * alpha_fore +
                    bg_band * alpha_back * (1 - alpha_fore)) / new_alpha_nan
```
are split in two steps: first compute the new value for the channels, and then apply the nans.

As this function is quite technical, maybe refactoring into even smaller function with relevant names could help clarify things?

Reply:
After some tests I suddenly realised that the whole NaN thing is completely unnecesary. By dropping it the logic is now almost the same as the old one and is easy to understand.

Reply:
That's great @yukaribbba , thanks a lot for looking into this again :) Looking forwards to your new commits